src/contentsign/ClickBounty.sol

```diff
@@ -0,0 +1,286 @@
+        if (amount > entryFee) {
+            token.safeTransfer(owner, amount - entryFee);
+        }
+
+        emit BountyIssued(tokenId, amount);
+
+        // Update the on-chain leaderboard
+        _updateLeaderboard(tokenId, amount);
+    }
+
+    // -------------------------------------------------------------------------
+    // View Functions
+    // -------------------------------------------------------------------------
+
+    /**
+     * @notice Retrieves an array of `(URI, bountyAmount)` for each token in the top winners list.
+     * @dev The size of the returned array is up to `LEADERBOARD_SIZE`, but might be smaller if
+     *      fewer than `LEADERBOARD_SIZE` bounties have been paid.
+     * @return uriBounties A dynamic array of `URIBounty` structs, each with `uri` and `bounty`.
+     */
+    function getLeaderboard() external view returns (URIBounty[] memory uriBounties) {
+        uriBounties = new URIBounty[](_leaderboard.length);
+        for (uint256 i = 0; i < _leaderboard.length; i++) {
+            uriBounties[i] = URIBounty({uri: contentSign.tokenURI(_leaderboard[i]), bounty: bounties[_leaderboard[i]]});
+        }
+    }
+
+    // -------------------------------------------------------------------------
+    // Internal Functions
+    // -------------------------------------------------------------------------
+
+    /**
+     * @dev Updates the leaderboard (`_leaderboard`) if `tokenId` is among the top bounties.
+     *      - If the leaderboard is not full, we simply append the new entry.
+     *      - Otherwise, we find the smallest bounty in `_leaderboard`; if `amount` exceeds that,
+     *        we replace it.
+     *
+     * @param tokenId The token ID that may enter the leaderboard.
+     * @param amount  The bounty amount to compare.
+     *
+     * Emits:
+     * - A {NewLeaderboardEntry} event if a replacement or addition occurs.
+     */
+    function _updateLeaderboard(uint256 tokenId, uint256 amount) private {
+        // If the leaderboard is not at capacity, just add the new entry
+        if (_leaderboard.length < LEADERBOARD_SIZE) {
+            _leaderboard.push(tokenId);
+            emit NewLeaderboardEntry(tokenId, amount);
+            return;
+        }
+
+        // Find the smallest bounty in the current leaderboard
+        uint256 smallestIndex = 0;
+        uint256 smallestAmount = bounties[_leaderboard[0]];
+        for (uint256 i = 1; i < _leaderboard.length; i++) {
+            uint256 currentAmount = bounties[_leaderboard[i]];
+            if (currentAmount < smallestAmount) {
+                smallestAmount = currentAmount;
+                smallestIndex = i;
+            }
+        }
+
+        // If the new amount is bigger than the smallest in the leaderboard, replace it
+        if (amount > smallestAmount) {
+            _leaderboard[smallestIndex] = tokenId;
+            emit NewLeaderboardEntry(tokenId, amount);
+        }
+    }
```

Comment:
What's the purpose of the leaderboard and why do we need this logic? I do not seem to find the leaderboard values used anywhere in the contract, which prompts me to think this would be better handled in an indexer than to spend gas on it.

Reply:
This could be done by an indexer. However since we only need a small board, the cost of keeping it on chain and updating it there is trivial. This means for our agile implementation we can quickly have it served from the chain. The additional advantage of having it on chain is that there would be a guaranteed minimal delay which is important for our use case. I will explain the use case on slack with you.